modules/platforms/cpp/ignite/client/detail/compute/job_execution_impl.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ignite/client/detail/compute/job_execution_impl.h"
+#include "ignite/client/detail/compute/compute_impl.h"
+
+namespace ignite::detail {
+
+void job_execution_impl::get_result_async(ignite_callback<std::optional<primitive>> callback) {
+    std::unique_lock<std::mutex> guard(m_mutex);
+
+    if (m_result) {
+        auto copy{*m_result};
+        guard.unlock();
+
+        callback({std::move(copy)});
+    } else if (m_error) {
+        auto copy{*m_error};
+        guard.unlock();
+
+        callback({std::move(copy)});
+    } else {
+        if (m_result_callback)
+            throw ignite_error("A callback for this result was already submitted");
```

Comment:
Please explain - when does this happen? As a user, can I call `get_result_async` multiple times?

Reply:
If it's called from multiple threads in parallel.

Reply:
This is a limitation, but I can not easily get rid of it  with the current approach to async execution.